### PR TITLE
Extract Aggregate, Sort, and Join to struct from AggregatePlan

### LINKED ISSUE #1303: Extract `Aggregate`, `Sort`, `Join` in LogicalPlan as independent struct
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
There are some discussions in #1228

**Describe the solution you'd like**
like #1290

**Describe alternatives you've considered**
The task is mechanical, however, it is ideal for newcomers to familiarize themselves with datafusion's build, test, and PR processes.

**Additional context**
If you want to do it, please ping me and assign the task to you, avoiding collisions with other people.

### PR DESCRIPTION
**Pull request by matthewmturner**
# Which issue does this PR close?

Closes #1303 
Closes https://github.com/apache/arrow-datafusion/issues/1228

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by xudong963** [comment]
Thanks @matthewmturner !

File: datafusion/src/logical_plan/plan.rs [hunk 1]
```diff
@@ -107,6 +107,19 @@ pub struct ExtensionPlan {
     pub node: Arc<dyn UserDefinedLogicalNode + Send + Sync>,
 }
 
+/// Aggregates its input based on a set of grouping and aggregate
+/// expressions (e.g. SUM).
+#[derive(Clone)]
+pub struct AggregatePlan {
```
Comment (alamb) [inline datafusion/src/logical_plan/plan.rs:113]: I think calling this `Aggregate` would be more consistent with what has been refactored in other PRs.

Let me know when this PR is ready @matthewmturner 

Thanks!

File: datafusion/src/logical_plan/plan.rs [hunk 2]
```diff
@@ -154,16 +167,17 @@ pub enum LogicalPlan {
...
-        /// The incoming logical plan
-        input: Arc<LogicalPlan>,
-        /// Grouping expressions
-        group_expr: Vec<Expr>,
-        /// Aggregate expressions
-        aggr_expr: Vec<Expr>,
-        /// The schema description of the aggregate output
-        schema: DFSchemaRef,
-    },
+    // Aggregate {
```
Comment (liukun4515) [inline datafusion/src/logical_plan/plan.rs:170]: in the final version, we should remove these useless code

File: datafusion/src/logical_plan/plan.rs [hunk 3]
```diff
@@ -289,7 +303,7 @@ impl LogicalPlan {
             LogicalPlan::Projection { schema, .. } => schema,
             LogicalPlan::Filter { input, .. } => input.schema(),
             LogicalPlan::Window { schema, .. } => schema,
-            LogicalPlan::Aggregate { schema, .. } => schema,
+            LogicalPlan::Aggregate(aggregate) => &aggregate.schema,
```
Comment (liukun4515) [inline datafusion/src/logical_plan/plan.rs:306]: @alamb I have question about the code style, before refactor, we use the style
```
::Aggregate {filed1, filed2, field3,....}
```
Someone may use the style after refactor

```
Aggregate(agg) => 
agg.field1
agg.field2
agg.field3
```
How about make consistent with before? 
In My pull requests #1325  #1322 #1311, I keep consistent and use [this](https://github.com/liukun4515/arrow-datafusion/blob/451a00cbe05ec385431eba97ba5a6d7a827fb982/ballista/rust/core/src/serde/logical_plan/to_proto.rs#L1000) style.

File: datafusion/src/logical_plan/plan.rs [hunk 4]
```diff
@@ -378,11 +392,12 @@ impl LogicalPlan {
                 _ => vec![],
             },
             LogicalPlan::Window { window_expr, .. } => window_expr.clone(),
-            LogicalPlan::Aggregate {
-                group_expr,
-                aggr_expr,
-                ..
-            } => group_expr.iter().chain(aggr_expr.iter()).cloned().collect(),
+            LogicalPlan::Aggregate(aggregate) => aggregate
```
Comment (liukun4515) [inline datafusion/src/logical_plan/plan.rs:395]: Same opinion like above, if we keep consistent like before, we can make less changes.

File: datafusion/src/optimizer/common_subexpr_eliminate.rs [hunk 5]
```diff
@@ -20,8 +20,8 @@
 use crate::error::Result;
 use crate::execution::context::ExecutionProps;
 use crate::logical_plan::{
-    col, DFField, DFSchema, Expr, ExprRewriter, ExpressionVisitor, LogicalPlan,
-    Recursion, RewriteRecursion,
+    col, plan::AggregatePlan, DFField, DFSchema, Expr, ExprRewriter, ExpressionVisitor,
```
Comment (liukun4515) [inline datafusion/src/optimizer/common_subexpr_eliminate.rs:23]: we can make the struct `AggregatePlan` pub in the logical plan package mod file

**Comment by alamb** [comment]
I think this PR is now the last remaining in the set for https://github.com/apache/arrow-datafusion/issues/1228 -- we are close now. @matthewmturner  any chance you can rebase this PR and we can wrap up this item?

**Comment by matthewmturner** [comment]
> I think this PR is now the last remaining in the set for #1228 -- we are close now. @matthewmturner any chance you can rebase this PR and we can wrap up this item?

yes, sry for delay. i am working on it.

i am going to align to style recommended by @liukun4515

**Comment by liukun4515** [comment]
> > I think this PR is now the last remaining in the set for #1228 -- we are close now. @matthewmturner any chance you can rebase this PR and we can wrap up this item?
> 
> yes, sry for delay. i am working on it.
> 
> i am going to align to style recommended by @liukun4515

If your pull request is ready, I will review it. 
@matthewmturner

**Comment by matthewmturner** [comment]
@liukun4515 ready for review!

File: datafusion/src/logical_plan/plan.rs (hunk 2 above)
Comment (liukun4515) [reply datafusion/src/logical_plan/plan.rs:313]: @matthewmturner

File: datafusion/src/logical_plan/plan.rs (hunk 4 above)
Comment (liukun4515) [reply datafusion/src/logical_plan/plan.rs:467]: please align the code style using
```
LogicalPlan::Aggregate(aggregate {
  field1，
  field2
})
```

**Review by liukun4515 — COMMENTED** [review]
LGTM, if you address the code style comments.
@matthewmturner

**Comment by xudong963** [comment]
LGTM, thanks @matthewmturner. You are the terminator of #1228 ! 🎉

**Review by houqp — APPROVED** [review]
LGTM!

**Review by alamb — APPROVED** [review]
Looking great -- thank you @matthewmturner @liukun4515  and @xudong963  for driving this refactoring

**Pull request merged**